client/state/data-layer/wpcom/jetpack-onboarding/index.js

```diff
@@ -116,13 +116,46 @@ export const saveJetpackOnboardingSettings = ( { dispatch, getState }, action )
 export const handleSaveSuccess = ( { dispatch }, { siteId, settings } ) =>
 	dispatch( saveJetpackOnboardingSettingsSuccess( siteId, settings ) );
 
-export const announceSaveFailure = ( { dispatch }, { siteId } ) =>
+export const announceSaveFailure = ( { dispatch }, { siteId } ) => {
 	dispatch(
 		errorNotice( translate( 'An unexpected error occurred. Please try again later.' ), {
 			id: `jpo-notice-error-${ siteId }`,
 			duration: 5000,
 		} )
 	);
+};
+
+export const retryOrAnnounceSaveFailure = ( { dispatch }, action ) => {
+	const MAX_WOOCOMMERCE_INSTALL_RETRIES = 2;
+	const { settings, siteId, type, meta: { dataLayer } } = action;
+	const { error, retryCount = 0 } = dataLayer;
+
+	// If we got a timeout on WooCommerce installation, try again (up to 3 times),
+	// since it might just be a slow server that actually ends up installing it
+	// properly, in which case a subsequent request will return 'success'.
+	if (
+		settings.installWooCommerce !== true ||
```

Comment:
could `settings` be `null` or `undefined` here?

Reply:
I agree with Dennis here; this could potentially happen in cases where we're not installing WooCommerce, so we might want to tighten this one.